aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,680 @@
+  int64_t* host_strides_ptr = host_strides.data_ptr<int64_t>();
+  if (ndim > 1) {
+    for (int64_t i = ndim - 2; i >= 0; i--) {
+      host_strides_ptr[i] =
+          host_strides_ptr[i + 1] * (i + 1 == dim ? 1 : sizes[i + 1]);
+    }
+  }
+  auto strides =
+      host_strides.to(at::Device(kCUDA), indices.dtype(), false, true);
+
+  auto strides_ptr = strides.data_ptr<int64_t>();
+
+  auto indices_accessor = indices.packed_accessor<int64_t, 2>();
+
+  Tensor offsets = at::empty({nnz}, indices.options());
+
+  thrust::transform(
+      policy,
+      thrust::make_counting_iterator(int64_t(0)),
+      thrust::make_counting_iterator(int64_t(nnz)),
+      thrust::device_ptr<int64_t>(offsets.data_ptr<int64_t>()),
+      [indices_accessor, strides_ptr, dim, ndim] __device__(int64_t x) {
+        int64_t pool_index = 0;
+        for (int64_t j = 0; j < ndim; j++) {
+          if (j != dim) {
+            auto indices_row = indices_accessor[j];
+            auto stride = strides_ptr[j];
+            pool_index += stride * indices_row[x];
+          }
+        }
+        return pool_index;
+      });
+  return offsets;
+}
+
+template <class scalar_t, bool requireMxRows = true>
+std::tuple<Tensor, Tensor, Tensor, Tensor> compute_pool_max(
+    const Tensor& indices,
+    const Tensor& values,
+    const IntArrayRef& sizes,
+    int64_t nvalues,
+    const int64_t dim) {
+  /*
+    Return pools of indices that align with the given dimension and the
+    correspoinding max values for each pool.
+
+    See ATen/native/sparse/cpu/Softmax.cpp:get_offsets and
+    ATen/native/sparse/cpu/Softmax.cpp:cpu_sparse_coo_softmax for the CPU
+    implementation that this implementation is based on.
+  */
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  auto nnz = indices.size(1);
+  auto offsets  = get_offsets(indices, sizes, dim);
+  int64_t* offsets_ptr = offsets.data_ptr<int64_t>();
+
+  auto sorted_indices = at::empty({nnz}, indices.options());
+  thrust_ptr sorted_indices_thrust_ptr(sorted_indices.data_ptr<int64_t>());
+  thrust::sequence(
+      policy, sorted_indices_thrust_ptr, sorted_indices_thrust_ptr + nnz, 0);
+
+  thrust::sort(
+      policy,
+      sorted_indices_thrust_ptr,
+      sorted_indices_thrust_ptr + nnz,
+      [offsets_ptr] __device__(int64_t x, int64_t y) {
+        return offsets_ptr[x] < offsets_ptr[y];
+      });
+  auto pool_sizes = at::empty({nnz}, indices.options());
+
+  auto new_end = thrust::reduce_by_key(
+      policy,
+      sorted_indices_thrust_ptr,
+      sorted_indices_thrust_ptr + nnz,
+      thrust::make_constant_iterator(int64_t(1)),
+      thrust::make_discard_iterator(),
+      thrust_ptr(pool_sizes.data_ptr<int64_t>()),
+      [offsets_ptr] __device__(int64_t x, int64_t y) {
+        return offsets_ptr[x] == offsets_ptr[y];
+      });
+  auto new_sz = thrust::distance(
```

Comment:
Can this just subtract the pointers to compute the difference instead of calling thrust::distance?

Reply:
It could be but with thrust the most common way to do that is using thrust::distance especially if we work with have to deal with `thrust::device_ptr`.